content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -7,4 +7,57 @@ next: /recipes/authoring/kata-snippets/preloaded/
+
+- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.
+- **Order of random tests should be non-deterministic.** Not only inputs should be randomized, but their order too. One possible technique is to generate and collect a set of random inputs for all required scenarios, and shuffle them.
+- If possible, **reference solution should not be used**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and it can save time which would be spent on running the reference solution.
+- **Assertions should be close to the calls of tested solution**, preferrably immediatelly following it. Debugging becomes difficult if invalid answer is verified long after it's returned.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent beetwen test runs.** When randomized ranges are very large, then it's possible that one user receives many small, easy inputs and their solution passes, while another user receives many large, more difficult inputs, and their solution fails or times out. Make sure that your random tests are built in a way which elimnates such situation. If you want to test difficult inputs, split test cases into a set of easy ones and a set of diffficult ones, and tets them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 
+
+## Performance Tests
+
+Some kata require solutions to be fast, or to fall into some performance category. For example, author might want to accept only solutions of complexity `O(n)`, and reject `O(n^2)` ones. Building such test suite is not always easy!
+
+- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
```

Comment:
```suggestion
- **Performance tests must be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
```

Reply:
Why "must"? It's not that random tests and performance tests always have to be one and the same block? Block of random tests **can** be used to test performance, but why would it always have to be?

Reply:
right :+1: